test/functional/legacy/061_undo_tree_spec.lua

```diff
@@ -0,0 +1,280 @@
+  helpers.clear, helpers.execute, helpers.expect, helpers.wait
+
+local function expect_empty_buffer()
+  -- The space will be removed by helpers.dedent but is needed as dedent will
+  -- throw an error if it can not find the common indent of the given lines.
+  return expect(' ')
+end
+local function expect_line(line)
+  return eq(line, eval('getline(".")'))
+end
+local function write_file(name, text)
+  local file = io.open(name, 'w')
+  file:write(text)
+  file:flush()
+  file:close()
+end
+
+describe('undo:', function()
+  before_each(clear)
+  teardown(function()
+    os.remove('Xtest.source')
+  end)
+
+  describe(':earlier and :later', function()
+    before_each(function()
+      os.remove('Xtest')
+    end)
+    teardown(function()
+      os.remove('Xtest')
+    end)
+
+    it('work with time specifications and g- and g+', function()
+      -- We write the test text to a file in order to prevent nvim to record
+      -- the inserting of the text into the undo history.
+      write_file('Xtest', '\n123456789\n')
+
+      -- Note: This test is not executed directly like the other tests.  It is
+      -- encapsuled in a function and will be tried up to three times if it
+      -- fails.  This is done because the test is dependent on wall clock time
+      -- and somtimes fails if the system is under load or the test is paused
+      -- for a long enough interval.
+
+      local function test()
+	execute('e Xtest')
+	-- Assert that no undo history is present.
+	eq({}, eval('undotree().entries'))
+	-- Delete three characters and undo.
+	feed('Gxxx')
+	expect_line('456789')
+	feed('g-')
+	expect_line('3456789')
+	feed('g-')
+	expect_line('23456789')
+	feed('g-')
+	expect_line('123456789')
+	feed('g-')
+	expect_line('123456789')
+
+	-- Delete three other characters and go back in time step by step.
+	feed('$xxx')
+	expect_line('123456')
+	execute('sleep 1')
+	wait()
+	feed('g-')
+	expect_line('1234567')
+	feed('g-')
+	expect_line('12345678')
+	feed('g-')
+	expect_line('456789')
+	feed('g-')
+	expect_line('3456789')
+	feed('g-')
+	expect_line('23456789')
+	feed('g-')
+	expect_line('123456789')
+	feed('g-')
+	expect_line('123456789')
+	feed('g-')
+	expect_line('123456789')
+	feed('10g+')
+	expect_line('123456')
+
+	-- Delay for two seconds and go some seconds forward and backward.
+	execute('sleep 2')
+	wait()
+	feed('Aa<esc>')
+	feed('Ab<esc>')
+	feed('Ac<esc>')
+	expect_line('123456abc')
+	execute('earlier 1s')
+	expect_line('123456')
+	execute('earlier 3s')
+	expect_line('123456789')
+	execute('later 1s')
+	expect_line('123456')
+	execute('later 1h')
+	expect_line('123456abc')
+      end
+      local success, result = false, ''
+      for i = 1, 3 do
+	success, result = pcall(test)
+	if success then
+	  return
+	end
+      end
+      -- We did not return in the loop so there was an error or failure.
+      -- We now try to run the test again but will not catch further errors,
+      -- so the user will see them.
```

Comment:
can just remove this comment and instead mention above line 104:

    Retry up to 3 times.